rust/rebase/src/flow/nft_ownership.rs

```diff
@@ -0,0 +1,241 @@
+                "Validation window has expired"
+            )));
+        };
+        Ok(())
+    }
+
+    pub async fn process_page(
+        &self,
+        client: &reqwest::Client,
+        u: url::Url,
+        contract_address: &str,
+    ) -> Result<PageResult, FlowError> {
+        let res: AlchemyNftRes = client
+            .get(u)
+            .send()
+            .await
+            .map_err(|e| FlowError::BadLookup(e.to_string()))?
+            .json()
+            .await
+            .map_err(|e| FlowError::BadLookup(e.to_string()))?;
+
+        let mut result: PageResult = PageResult {
+            next_page: res.page_key.clone(),
+            found: false,
+            // res: res.clone(),
+        };
+
+        for nft in res.owned_nfts {
+            if nft.contract.address == contract_address {
+                result.found = true;
+                break;
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+#[async_trait(?Send)]
+impl Flow<Ctnt, Stmt, Prf> for NftOwnership {
+    fn instructions(&self) -> Result<Instructions, FlowError> {
+        Ok(Instructions {
+            statement: "Enter the contract address and network of asset".to_string(),
+            signature: "Sign a statement attesting to ownership of the asset".to_string(),
+            witness: "Send the attestation and the signature to the witness and issue a credential"
+                .to_string(),
+            statement_schema: schema_for!(Stmt),
+            witness_schema: schema_for!(Prf),
+        })
+    }
+
+    async fn statement<I: Issuer>(
+        &self,
+        stmt: &Stmt,
+        _issuer: &I,
+    ) -> Result<FlowResponse, FlowError> {
+        self.sanity_check(&stmt.issued_at)?;
+
+        // TODO: Adjust this when adding additional Alchemy flows.
+        match stmt.subject {
+            Subjects::Pkh(Pkh::Eip155(_)) => {}
+            _ => {
+                return Err(FlowError::Validation(
+                    "Currently only supports Ethereum NFTs".to_string(),
+                ))
+            }
+        }
+
+        Ok(FlowResponse {
+            statement: stmt.generate_statement()?,
+            delimitor: None,
+        })
+    }
+
+    async fn validate_proof<I: Issuer>(&self, proof: &Prf, _issuer: &I) -> Result<Ctnt, FlowError> {
+        self.sanity_check(&proof.statement.issued_at)?;
+
+        let u = Url::parse(&format!(
+            "https://{}-{}.g.alchemy.com/nft/v2/{}/getNFTs?owner={}&withMetadata=false",
+            // TODO: Replace with enum.
+            "eth".to_string(),
+            // TODO: Replace with enum.
+            proof.statement.network,
+            self.api_key,
+            proof.statement.subject.display_id()?
+        ))
+        .map_err(|e| FlowError::BadLookup(e.to_string()))?;
+
+        let client = Client::new();
+        // NOTE: Intentionally re-assigning before read, thus using: #[allow(unused_assignments)]
+        #[allow(unused_assignments)]
```

Comment:
I don't understand

Reply:
I want to modify them in the loop. I re-assign based on the first iteration.

Reply:
`page_key` is assigned and then directly used, so you don't need it to be "global", you can just do 
```rust
let page_key = res.next_page.clone();
```

Reply:
I'll move `res` and `page_key` into the loop? What about `next_u`, how do we get around that?

Reply:
`res` is used after the loop, how do I do that? I tried returning from the loop, but that seems to return out of the function:
```

        let res = loop {
            let inner_res = self
                .process_page(&client, next_u.clone(), &proof.statement.contract_address)
                .await?;

            if inner_res.found {
                return inner_res;
            }

            let page_key = inner_res.next_page.clone();
            match page_key {
                None => return inner_res,
                Some(s) => {
                    next_u = u.join(&format!("&pageKey={}", s)).map_err(|_e| {
                        FlowError::BadLookup("Could not follow paginated results".to_string())
                    })?;
                }
            }
        };
  ```

Reply:
You can do `break inner_res`